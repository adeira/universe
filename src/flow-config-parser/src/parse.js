// @flow strict

import { isNumeric, invariant } from '@adeira/js';

import { isListSection, KNOWN_LIST_OPTIONS } from './utils';
import type { ParsedConfig } from './ParsedConfig.flow';

export default function parse(input: string): ParsedConfig {
  const lines = input
    .split(/\r?\n+/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  const parsedConfig = {
    // default values (TODO: remove and do not return defaults?)
    declarations: [],
    ignore: [],
    include: [],
    libs: [],
    lints: null,
    options: null,
    rollouts: null,
    strict: [],
    untyped: [],
    version: null,
  };

  let currentSection = null;
  for (const line of lines) {
    const commentMatch = line.match(/[;#ðŸ’©]/u);
    if (commentMatch) {
      continue;
    }

    const sectionMatch = line.match(/\[(?<sectionName>[a-z]+)]/i);
    if (sectionMatch) {
      // TODO: validate sections (?)
      currentSection = sectionMatch.groups?.sectionName.toLowerCase();
      continue;
    }

    invariant(currentSection != null, 'All config values must be inside some section.');

    if (isListSection(currentSection)) {
      // process it as a list
      parsedConfig[currentSection].push(line);
    } else if (currentSection === 'version') {
      /* $FlowFixMe[incompatible-type] This comment suppresses an error when
       * upgrading Flow. To see the error delete this comment and run Flow. */
      parsedConfig[currentSection] = line;
    } else {
      // process it as a key/value (while ignoring rollout annotations since we do not merge them properly, yet)
      const keyValueMatch = line.match(/^(?:(?<rollout>\(.+\)))?(?<rawKey>.+)=(?<rawValue>.+)$/);
      if (keyValueMatch) {
        const rawKey = keyValueMatch.groups?.rawKey;
        const rawValue = keyValueMatch.groups?.rawValue;
        const key = rawKey?.trim();
        if (parsedConfig[currentSection] === null) {
          /* $FlowFixMe[prop-missing] This comment suppresses an error when
           * upgrading Flow to version 0.191.0. To see the error delete this
           * comment and run Flow. */
          parsedConfig[currentSection] = {};
        }
        if (KNOWN_LIST_OPTIONS.includes(key)) {
          const previousValues = parsedConfig[currentSection][key] ?? [];
          parsedConfig[currentSection][key] = previousValues.concat(rawValue);
        } else {
          parsedConfig[currentSection][key] = transformKeyValue(rawValue);
        }
      }
    }
  }

  return parsedConfig;
}

function transformKeyValue(value: void | string) {
  if (value?.toLowerCase() === 'true') {
    return true;
  } else if (value?.toLowerCase() === 'false') {
    return false;
  } else if (isNumeric(value)) {
    return Number(value);
  }
  return value;
}
