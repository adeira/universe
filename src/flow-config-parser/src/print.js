// @flow strict

import { KNOWN_LIST_OPTIONS } from './utils';
import type { ParsedConfig } from './ParsedConfig.flow';

export default function print(flowconfig: ParsedConfig): string {
  let printedConfig = '';

  let sep = '';
  for (const sectionName of Object.keys(flowconfig)) {
    const sectionValues = flowconfig[sectionName];
    printedConfig += `${sep}[${sectionName}]\n`;
    sep = '\n';

    if (Array.isArray(sectionValues)) {
      printedConfig += `${sectionValues.join('\n')}\n`;
    } else if (sectionName === 'version' && sectionValues != null) {
      printedConfig += `${sectionValues}\n`;
    } else if (sectionValues != null) {
      // go through the individual values and print them
      for (const sectionValueKey of Object.keys(sectionValues)) {
        const sectionValue = sectionValues[sectionValueKey];
        if (KNOWN_LIST_OPTIONS.includes(sectionValueKey)) {
          for (const value of sectionValue) {
            printedConfig += `${sectionValueKey}=${value}\n`;
          }
        } else {
          printedConfig += `${sectionValueKey}=${sectionValue}\n`;
        }
      }
    }
  }

  return printedConfig;
}
