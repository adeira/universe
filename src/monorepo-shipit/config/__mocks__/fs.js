// @flow

const fs: any = jest.createMockFromModule('fs');

fs.existsSync = (path) => path !== '/unknown_path';

export default fs;
