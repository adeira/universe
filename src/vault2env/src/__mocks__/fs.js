// @flow

const fs: any = jest.createMockFromModule('fs');

let mockFiles = Object.create(null);
fs.__setMockFiles = function (newMockFiles) {
  mockFiles = newMockFiles;
};

fs.existsSync = function (filePath) {
  return mockFiles[filePath] !== undefined;
};

fs.writeFileSync = function (filePath, content) {
  mockFiles[filePath] = content;
};

fs.readFileSync = function (filePath) {
  return mockFiles[filePath];
};

module.exports = fs;
