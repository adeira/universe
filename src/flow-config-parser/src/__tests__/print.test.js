// @flow strict

import fs from 'fs';
import path from 'path';

import parse from '../parse';
import print from '../print';

it('prints the config correctly', () => {
  const parsedConfig = parse(
    fs.readFileSync(path.join(__dirname, 'fixtures', '.flowconfig.universe'), 'utf8'),
  );
  const printedConfig = print(parsedConfig);

  expect(printedConfig).toMatchSnapshot();

  // let's parse the printed config once again and compare it with the original (it should match)
  expect(parse(printedConfig)).toStrictEqual(parsedConfig);
});
