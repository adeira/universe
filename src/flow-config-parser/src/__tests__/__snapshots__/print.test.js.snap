// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prints the config correctly 1`] = `
"[declarations]
.+/node_modules/.+

[ignore]
<PROJECT_ROOT>/\\.git/.*
.+/.+\\.gzip\\.js
.+/\\.next/.+
.+/bazel-bin/.+
.+/bazel-out/.+
.+/bazel-testlogs/.+
.+/bazel-universe/.+

[include]


[libs]
./flow-typed/
./src/monorepo-scanner/src/flow/

[lints]
all=error
ambiguous-object-type=off
sketchy-null-bool=off
unclear-type=off
untyped-import=off

[options]
emoji=true
exact_by_default=true
include_warnings=true
facebook.fbt=FbtWithoutString
react.runtime=automatic
module.ignore_non_literal_requires=true
suppress_type=$FlowFixMe
experimental.const_params=true

[rollouts]

[strict]
ambiguous-object-type
sketchy-null-bool
unclear-type
untyped-import

[untyped]


[version]
"
`;
